File: tst/id.c
#include "pico.h"
#include <stdio.h>
#include <string.h>

int main (void) {
    pico_init(1);

    printf("=== pico_id: monotonic ===\n");
    {
        int a = pico_id();
        int b = pico_id();
        assert(a >= 1);
        assert(b == a+1);
    }

    printf("=== constructors return the stored key ===\n");
    {
        const char* k = pico_layer_empty("world", "e1", 1, (Pico_Rel_Dim){'!', {50, 50}}, NULL);
        assert(strcmp(k, "e1") == 0);
        const char* old = pico_set_layer(k);
        assert(strcmp(pico_get_layer(), "e1") == 0);
        pico_set_layer(old);
    }
    {
        const char* k = pico_layer_text(NULL, "t1", (Pico_Rel_Dim){'!', {0, 20}}, "hi");
        assert(strcmp(k, "t1") == 0);
    }
    {
        static Pico_Color buf[4] = {
            {255, 0, 0, 255}, {0, 255, 0, 255},
            {0, 0, 255, 255}, {255, 255, 0, 255}
        };
        const char* k = pico_layer_pixmap(NULL, "p1", (Pico_Abs_Dim){2, 2}, buf);
        assert(strcmp(k, "p1") == 0);
    }

    printf("=== key==NULL: auto-generates /id-N ===\n");
    {
        const char* a = pico_layer_empty(NULL, NULL, 0, (Pico_Rel_Dim){'!', {10, 10}}, NULL);
        const char* b = pico_layer_empty(NULL, NULL, 0, (Pico_Rel_Dim){'!', {10, 10}}, NULL);
        assert(strncmp(a, "/id-", 4) == 0);
        assert(strncmp(b, "/id-", 4) == 0);
        assert(strcmp(a, b) != 0);

        const char* s = pico_layer_sub(NULL, NULL, a, &(Pico_Rel_Rect){'%', {.5, .5, 1, 1}, PICO_ANCHOR_C});
        assert(strncmp(s, "/id-", 4) == 0);

        const char* h = pico_layer_screenshot(NULL, NULL, a, NULL);
        assert(strncmp(h, "/id-", 4) == 0);

        const char* t = pico_layer_text(NULL, NULL, (Pico_Rel_Dim){'!', {0, 20}}, "auto");
        assert(strncmp(t, "/id-", 4) == 0);
    }

    printf("=== image: key==NULL defaults to path ===\n");
    {
        const char* k = pico_layer_image(NULL, NULL, "../res/open.png");
        assert(strcmp(k, "../res/open.png") == 0);
    }

    printf("\n=== ALL TESTS PASSED ===\n");

    pico_init(0);
    return 0;
}

File: lua/tst/id.lua
pico.init(true)

print "pico.id: monotonic integer"
do
    local a = pico.id()
    local b = pico.id()
    assert(math.type(a) == 'integer')
    assert(a >= 1)
    assert(b == a+1)
end

print "pico.id: prefix form"
do
    local s = pico.id("bullet")
    local n = tonumber(s:match("^bullet%-(%d+)$"))
    assert(n ~= nil)
    assert(pico.id() == n+1)
end

print "layer constructors return the key"
do
    local e = pico.layer.empty { up="world", key="e1", clear=true, dim={'!', w=50, h=50} }
    assert(e == "e1")
    local t = pico.layer.text { key="t1", dim={'!', h=20}, text="hi" }
    assert(t == "t1")
    local p = pico.layer.pixmap { key="p1", pixels={{{r=255, g=0, b=0, a=255}}} }
    assert(p == "p1")
end

print "key omitted: auto-generates /id-N"
do
    local a = pico.layer.empty { dim={'!', w=10, h=10} }
    local b = pico.layer.empty { dim={'!', w=10, h=10} }
    assert(a:match("^/id%-%d+$"))
    assert(b:match("^/id%-%d+$"))
    assert(a ~= b)
    local s = pico.layer.sub { sup=a, crop={'%', x=0.5, y=0.5, w=1, h=1} }
    assert(s:match("^/id%-%d+$"))
    local h = pico.layer.screenshot { sup=a }
    assert(h:match("^/id%-%d+$"))
end

print "image: key omitted defaults to path"
do
    local k = pico.layer.image { path="../../res/open.png" }
    assert(k == "../../res/open.png")
end

print "\n=== ALL TESTS PASSED ==="
pico.init(false)
